### PR TITLE
ChibiOS: added memory guard system

### PR DESCRIPTION
**Pull request by tridge**
This is a system for finding memory corruption bugs. Enable using --enable-memory-guard

### CODE REVIEW

**Comment by tridge** [comment]
this system was used to find #15807

**Review by peterbarker — COMMENTED** [review]
Didn't spot anything wrong, but @bugobliterator definitely should look.

File: libraries/AP_HAL_ChibiOS/Scheduler.cpp [hunk 1]
```diff
@@ -492,6 +492,25 @@ void Scheduler::_io_thread(void* arg)
     }
 }
 
+#if defined(STM32H7)
+/*
+  the H7 has 64k of ITCM memory at address zero. We reserve 1k of it
+  to prevent nullptr being valid. This function checks that memory has
+  not been written to
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Scheduler.cpp]: fixed

File: libraries/AP_HAL_ChibiOS/Scheduler.h [hunk 2]
```diff
@@ -172,5 +172,9 @@ class ChibiOS::Scheduler : public AP_HAL::Scheduler {
     void _run_timers();
     void _run_io(void);
     static void thread_create_trampoline(void *ctx);
+
+#if defined STM32H7
+    void check_low_memory();
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Scheduler.h]: fixed

File: libraries/AP_HAL_ChibiOS/hwdef/common/malloc.c [hunk 3]
```diff
@@ -71,6 +80,19 @@ static memory_heap_t dma_reserve_heap;
...
+    chMtxObjectInit(&mem_mutex);
+#endif
+
+#if defined(STM32H7)
+    // zero first 1k of ITCM. We leave 1k free to avoid addresses
+    // close to nullptr being valid. Zeroing it here means we can
+    // check for changes which indicate a write to an uninitialised
+    // object.  We start at address 0x4 as writing the first word
+    // causes a fault
+    memset((void*)0x00000004, 0, 1020);
```
Comment (bugobliterator) [inline libraries/AP_HAL_ChibiOS/hwdef/common/malloc.c:93]: So this is only able to track down issues like writing to a ptr near 0x0 but just not at 0x0 itself.

Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/hwdef/common/malloc.c:93]: Also is it valid to set value at 0x0 in H7 board?

Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/common/malloc.c:93]: i get a hard fault if I try to write at 0x0. I vaguely remember there is a control register that enables/disables access to address 0, but I can't find it at the moment.

Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/common/malloc.c]: it turns out we can read/write all but the first byte. I've changed the code to check address 0x1 to 0x3 as well

**Review by peterbarker — APPROVED** [review]

**Review by bugobliterator — APPROVED** [review]

**Comment by peterbarker** [comment]
Merged, thanks.

**Pull request merged**